frontend/src/AnnotatedText.tsx

```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import * as firebase from "firebase/app";
+import {MedicalAnnotation, AnnotationEntity} from "./MedicalAnnotation"
+
+
+interface AnnotationResult {
+    data: {
+        Entities: Array<AnnotationEntity>
+    }
+}
+
+interface AnnotationProps {
+    message: string
+}
+
+interface AnnotationCache {
+    [indexer: string] : AnnotationResult
+}
+const cache: AnnotationCache = {}
+
+export default function AnnotatedText(props: AnnotationProps)  {
+    const [result, setResult] = React.useState<AnnotationResult | null>(null);
+
+    React.useEffect(() => {
+        let annotateTranscription = firebase.functions().httpsCallable('annotateTranscription');
+        if (props.message in cache) {
+            console.log("Cached!");
+            setResult(cache[props.message]);
+        } else {
+            annotateTranscription({'message': props.message})
+                .then((response: AnnotationResult) => {
+                    cache[props.message] = response;
+                    setResult(response);
+                });
+        }
+    });
+
+    let last = 0;
+    if (result) {
+        console.log("Result length: " + props.message.length);
```

Comment:
Consider a ternary operator.

Reply:
I find ternary's really are best for short things, not large blocks.